delira/training/predictor.py

```diff
@@ -0,0 +1,208 @@
+
+    def _setup(self, network, convert_batch_to_npy_fn, prepare_batch_fn):
+        self.module = network
+        self._convert_batch_to_npy_fn = convert_batch_to_npy_fn
+        self._prepare_batch = prepare_batch_fn
+
+    def __call__(self, data):
+        return self.predict(data)
+
+    def predict(self, data, keys=["data"]):
+        data = self._prepare_batch(data)
+
+        if isinstance(data, dict):
+            data = [data[key] for key in keys]
+
+        return self._convert_batch_to_npy_fn(
+            self.module(
+                *data
+            )
+        )
+
+    def predict_data_mgr(self, datamgr, batchsize=None, verbose=False):
+        """
+        Defines a routine to predict data obtained from a batchgenerator
+
+        Parameters
+        ----------
+        batchgen : MultiThreadedAugmenter
+            Generator Holding the Batches
+        batchsize : Artificial batchsize (sampling will be done with batchsize
+                    1 and sampled data will be stacked to match the artificial
+                    batchsize)(default: None)
+
+        Raises
+        ------
+        NotImplementedError
+            If not overwritten by subclass
+
+        """
+
+        orig_num_aug_processes = datamgr.n_process_augmentation
+        orig_batch_size = datamgr.batch_size
+
+        datamgr.batch_size = 1
+        datamgr.n_process_augmentation = 1
+
+        batchgen = datamgr.get_batchgen()
+
+        predictions_all, inputs_all = [], []
+
+        n_batches = batchgen.generator.num_batches * batchgen.num_processes
+
+        if verbose:
+            iterable = tqdm(enumerate(batchgen), unit=' sample',
+                            total=n_batches, desc='Test')
+
+        else:
+            iterable = enumerate(batchgen)
+
+        batch_list = []
+
+        for i, batch in iterable:
+
+            if not batch_list and (n_batches - i) < batchsize:
+                batchsize = n_batches - i
+                logger.debug("Set Batchsize down to %d to avoid cutting "
+                             "of the last batches" % batchsize)
+
+            batch_list.append(batch)
+
+            # if queue is full process queue:
+            if batchsize is None or len(batch_list) >= batchsize:
+
+                batch_dict = {}
+                for batch in batch_list:
+                    for key, val in batch.items():
+                        if key in batch_dict.keys():
+                            batch_dict[key].append(val)
+                        else:
+                            batch_dict[key] = [val]
+
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.array(val_list)
+
+                preds = self.predict(batch_dict)
+
+                predictions_all.append(preds)
+                inputs_all.append(batch_dict)
+
+                batch_list = []
+
+        batchgen._finish()
+
+        predictions_all = zip(*predictions_all)
```

Comment:
For `zip` to work properly it is necessary that `predictions_all` is a list of lists. For some networks (like detection networks) which complete multiple tasks (detection, segmentation, classification) I like to use a dict for my network output (so I don't have to remember which entry belongs to which task). Is there any possibility to allow that here?

Reply:
We would have to entirely switch to a dict, because maintaining both possibilities would be an unnecessary overhead I guess.

What do you think @haarburger @ORippler ?

Reply:
I think we shouldn't mix `lists` and `dicts`. In current master, we only use lists for `labels_all` and `outputs_all`, but now we intermingle it already now with `predictions_all` and `total_input_dict`. If we need the whole information in `total_input_dict` in order to calculate proper validation metrics, we should switch to dicts totally.

Reply:
I think, I'd like to move this to another issue, since this issue shouldn't change anything on the user-side, but switching from lists to dicts will.